{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Integrated Bioinformatics Project : Flux_Balance_F_Prausnitzii \n",
    "#### Students: Michael Shawn Neilsson, Biancamaria FLorenzi, Prabhat Juyal, Tim Blokker\n",
    "#### Supervisors: ClÃ©mence Joseph, Karoline Faust"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#import packages\n",
    "import io\n",
    "import cobra\n",
    "import functions\n",
    "from functions import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#model = cobra.io.read_sbml_model('Data/f_prausnitzii.xml')\n",
    "#info(model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Metabolites :  833\n",
      "Reactions :  1030\n",
      "Genes :  602\n",
      "Compartments :  {'c': '', 'e': ''}\n",
      "Objective function : \n",
      " Maximize\n",
      "1.0*Biomass_FP - 1.0*Biomass_FP_reverse_ee33b \n",
      "\n"
     ]
    }
   ],
   "source": [
    "model=cobra.io.load_matlab_model('Data/iFpraus_v_1_0.mat')\n",
    "info(model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#https://www.thielelab.eu/in-silico-models\n",
    "#model = cobra.io.read_sbml_model('Data/iFpraus_v_1_0.xml')\n",
    "#(model2,errors)=(cobra.io.sbml.validate_sbml_model('Data/iFpraus_v_1_0.xml'))\n",
    "#print(errors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "outputmodel= cobra.io.save_json_model(model, \"Data/iFpraus.json\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#use for later perhaps \n",
    "#modelMedium = setMediumConstraints(starting_model, set_inf,\n",
    "#current_inf,... medium_composition, met_Conc_mM, cellConc, t, cell-\n",
    "#Weight, mediumCompounds,... mediumCompounds_lb);"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What are the different reactions\n",
    "e: extracellular\n",
    "\n",
    "c: cytosol\n",
    "\n",
    "EX_ : exchange reaction\n",
    "### Gifu Anaerobic Medium (mGAM) from:\n",
    "https://hyserve.com/files/05433_GAM-Broth_Modified_final.pdf\n",
    "\n",
    "\n",
    "- Peptone, Soya Peptone, Proteose Peptone: source of amino acids -> most likely all of them \n",
    "- Yeast extract, liver extract, meat extract, digested serum -> amino acids but also a lot of other crap, will leave this till the end \n",
    "- Dextrose -> same as glucose (https://www.ebi.ac.uk/chebi/searchId.do?chebiId=CHEBI:17634) and so -> glc_D\n",
    "- Soluble starch -> starch1200 (potatoe starch)\n",
    "- L-Tryptophane -> trp__L\n",
    "- L-Cystein Hydrochloride -> cys_L and cl and h (hydrochloride<-> HCL becomes cl and h)\n",
    "- Sodium Thioglycolate -> thiog and na1\n",
    "- L-Arginine -> arg_L\n",
    "- Vitamine K1 -> phllqne https://pubchem.ncbi.nlm.nih.gov/compound/Vitamin-K1#section=Depositor-Supplied-Synonyms \n",
    "- Hemin -> Iron and Chlorine  fe3 and cl  and perhaps ppp9 https://pubchem.ncbi.nlm.nih.gov/#query=3-%5B18-(2-carboxyethyl)-8%2C13-bis(ethenyl)-3%2C7%2C12%2C17-tetramethylporphyrin-21%2C23-diid-2-yl%5Dpropanoic%20acid%3Biron(3%2B)%3Bchloride \n",
    "- Potassium Dihydrogen Phosphate -> pi, k, h\n",
    "- Sodium Chloride -> na1 cl\n",
    "\n",
    "Digested serum : https://www.nebiolabs.com.au/-/media/catalog/datacards-or-manuals/p8108datasheet-lot0021306.pdf\t\n",
    "\n",
    "Generally useful page ; https://pubchem.ncbi.nlm.nih.gov/#query=Hemin\n",
    "\n",
    "check also: https://opencobra.github.io/cobratoolbox/latest/tutorials/tutorialMetabotoolsI.html "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ala_L(e)', 'arg_L(e)', 'asn_L(e)', 'asp_L(e)', 'asx_L(e)', 'cys_L(e)', 'glu_L(e)', 'gln_L(e)', 'glx_L(e)', 'gly_L(e)', 'his_L(e)', 'ile_L(e)', 'leu_L(e)', 'lys_L(e)', 'met_L(e)', 'phe_L(e)', 'pro_L(e)', 'ser_L(e)', 'thr_L(e)', 'trp_L(e)', 'tyr_L(e)', 'val_L(e)']\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "aa=(\"ala,arg,asn,asp,asx,cys,glu,gln,glx,gly,his,ile,leu,lys,met,phe,pro,ser,thr,trp,tyr,val\").split(\",\")\n",
    "aa=[aminoacid +\"_L(e)\" for aminoacid in aa]\n",
    "print(aa)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Model Medium\n",
      "Objective\n",
      "=========\n",
      "1.0 Biomass_FP = 69.87885590787737\n",
      "\n",
      "Uptake\n",
      "------\n",
      "    Metabolite          Reaction   Flux  C-Number  C-Flux\n",
      "    Lcystin[e]     EX_Lcystin(e)  3.502         6   0.05%\n",
      "        ade[e]         EX_ade(e)  864.9         5  10.85%\n",
      "     adocbl[e]      EX_adocbl(e) 0.3929        72   0.07%\n",
      "     alaala[e]      EX_alaala(e)  6.462         6   0.10%\n",
      "     alagln[e]      EX_alagln(e)  30.81         8   0.62%\n",
      "      arg-L[e]       EX_arg_L(e)   1000         6  15.05%\n",
      "      asn-L[e]       EX_asn_L(e)   16.4         4   0.16%\n",
      "        btn[e]         EX_btn(e) 0.3929        10   0.01%\n",
      "        ca2[e]         EX_ca2(e) 0.3929         0   0.00%\n",
      "         cl[e]          EX_cl(e) 0.3929         0   0.00%\n",
      "    cobalt2[e]     EX_cobalt2(e) 0.3929         0   0.00%\n",
      "     cps_fp[e]      EX_cps_fp(e) 0.3929        63   0.06%\n",
      "        cu2[e]         EX_cu2(e) 0.3929         0   0.00%\n",
      "        fe3[e]         EX_fe3(e) 0.7858         0   0.00%\n",
      "        fol[e]         EX_fol(e)  1.179        19   0.06%\n",
      "        fum[e]         EX_fum(e)   1000         4  10.03%\n",
      "      glu-L[e]       EX_glu_L(e)   1000         5  12.54%\n",
      "        gua[e]         EX_gua(e)  16.25         5   0.20%\n",
      "          h[e]           EX_h(e)    901         0   0.00%\n",
      "      his-L[e]       EX_his_L(e)  6.472         6   0.10%\n",
      "        ins[e]         EX_ins(e)  751.8        10  18.86%\n",
      "          k[e]           EX_k(e) 0.3929         0   0.00%\n",
      "     leuleu[e]      EX_leuleu(e)  15.33        12   0.46%\n",
      "      lys-L[e]       EX_lys_L(e)  23.31         6   0.35%\n",
      "      mal-L[e]       EX_mal_L(e)   1000         4  10.03%\n",
      "     metala[e]      EX_metala(e)  10.59         8   0.21%\n",
      " metsox-S-L[e]  EX_metsox_S_L(e) 0.2368         5   0.00%\n",
      "        mg2[e]         EX_mg2(e) 0.3929         0   0.00%\n",
      "        na1[e]         EX_na1(e)   1000         0   0.00%\n",
      "        nac[e]         EX_nac(e) 0.7858         6   0.01%\n",
      "         o2[e]          EX_o2(e)  618.9         0   0.00%\n",
      "      phe-L[e]       EX_phe_L(e)  12.65         9   0.29%\n",
      "         pi[e]          EX_pi(e)  126.6         0   0.00%\n",
      "     pnto-R[e]      EX_pnto_R(e) 0.7858         9   0.02%\n",
      "      pro-L[e]       EX_pro_L(e)  15.05         5   0.19%\n",
      "       pydx[e]        EX_pydx(e) 0.3929         8   0.01%\n",
      "     ribflv[e]      EX_ribflv(e)  148.3        17   6.32%\n",
      "      ser-L[e]       EX_ser_L(e)  223.1         3   1.68%\n",
      "        so4[e]         EX_so4(e) 0.3929         0   0.00%\n",
      "        thm[e]         EX_thm(e) 0.3929        12   0.01%\n",
      "      thr-L[e]       EX_thr_L(e)  606.4         4   6.08%\n",
      "      thymd[e]       EX_thymd(e)  1.642        10   0.04%\n",
      "      trp-L[e]       EX_trp_L(e)  3.854        11   0.11%\n",
      "      tyr-L[e]       EX_tyr_L(e)  9.381         9   0.21%\n",
      "        ura[e]         EX_ura(e)  20.52         4   0.21%\n",
      "      urate[e]       EX_urate(e)  398.4         5   5.00%\n",
      "       hpyr[c]         sink_hpyr   0.05         3   0.00%\n",
      "\n",
      "Secretion\n",
      "---------\n",
      " Metabolite       Reaction   Flux  C-Number  C-Flux\n",
      "     na1[c]         DM_NA1  -1000         0   0.00%\n",
      "      ac[e]       EX_ac(e)  -1000         2   5.64%\n",
      "   asp-L[e]    EX_asp_L(e) -977.1         4  11.02%\n",
      "     cit[e]      EX_cit(e) -988.6         6  16.73%\n",
      "     co2[e]      EX_co2(e)  -1000         1   2.82%\n",
      "  dhor-S[e]   EX_dhor_S(e) -785.4         5  11.07%\n",
      "     for[e]      EX_for(e)   -158         1   0.45%\n",
      "     gly[e]      EX_gly(e) -527.8         2   2.98%\n",
      "  glyc-R[e]   EX_glyc_R(e)  -0.05         3   0.00%\n",
      "    hxan[e]     EX_hxan(e)  -1000         5  14.10%\n",
      "     nh4[e]      EX_nh4(e)  -1000         0   0.00%\n",
      "    orot[e]     EX_orot(e) -194.5         5   2.74%\n",
      "    ptrc[e]     EX_ptrc(e) -979.9         4  11.05%\n",
      " rbflvrd[e]  EX_rbflvrd(e) -147.5        17   7.07%\n",
      "    succ[e]     EX_succ(e) -20.14         4   0.23%\n",
      "     xan[e]      EX_xan(e)  -1000         5  14.10%\n",
      "\n"
     ]
    }
   ],
   "source": [
    "model1=model.copy()\n",
    "print(\"\\nModel Medium\")\n",
    "#model1.objective = {model1.reactions.get_by_id('EX_ac(e)'): 1}\n",
    "model1.objective = {model1.reactions.get_by_id('Biomass_FP'): 1}\n",
    "\n",
    "print(model1.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#met_Conc_mM = [0.1;1.15;0.15;0.379;0.208;2;0.136;0.133;0.0968;0.382;0.382;0.274;0.101;0.0909;0.153;0.174;...\n",
    "#0.286;0.168;0.0245;0.129;0.171;0.00863;0.00082;0.0214;0.000524;0.00227;0.082;0.00485;0.000532;0.00297;...\n",
    "#0.194;0.424;0;5.33;23.81;127.26;5.63;11.11;0;0;0;1;0;0.00326;0.0073]\n",
    "#customizedConstraints_lb = [-2.3460;0;0;-500;-100;-100;-100];\n",
    "#customizedConstraints_ub = [500;0;0;500;500;500;500];"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ala_L(e)', 'arg_L(e)', 'asn_L(e)', 'asp_L(e)', 'asx_L(e)', 'cys_L(e)', 'glu_L(e)', 'gln_L(e)', 'glx_L(e)', 'gly_L(e)', 'his_L(e)', 'ile_L(e)', 'leu_L(e)', 'lys_L(e)', 'met_L(e)', 'phe_L(e)', 'pro_L(e)', 'ser_L(e)', 'thr_L(e)', 'trp_L(e)', 'tyr_L(e)', 'val_L(e)', 'glc_D(e)', 'starch1200(e)', 'cl(e)', 'h(e)', 'ac(e)', 'so4(e)', 'na1(e)', 'fe3(e)', 'ppp9(e)', 'h2o(e)', 'pi(e)', 'k(e)']\n",
      "\n",
      "Medium_mgam\n",
      "Objective\n",
      "=========\n",
      "1.0 Biomass_FP = 4.3671073090673347e-17\n",
      "\n",
      "Uptake\n",
      "------\n",
      "Metabolite   Reaction  Flux  C-Number   C-Flux\n",
      "    na1[e]  EX_na1(e) 33.47         0    0.00%\n",
      "   hpyr[c]  sink_hpyr  0.05         3  100.00%\n",
      "\n",
      "Secretion\n",
      "---------\n",
      "Metabolite   Reaction   Flux  C-Number  C-Flux\n",
      "    na1[c]     DM_NA1 -33.47         0   0.00%\n",
      "     ac[e]   EX_ac(e)  -0.05         2  66.67%\n",
      "    co2[e]  EX_co2(e)  -0.05         1  33.33%\n",
      "\n"
     ]
    }
   ],
   "source": [
    "MediumMgam = ['ala_L_(e)', 'arg_L_(e)', 'asn_L_(e)', 'asp_L_(e)', 'asx_L_(e)', 'cys_L_(e)', 'glu_L_(e)', 'gln_L_(e)', \n",
    "              'glx_L_(e)', 'gly_L_(e)', 'his_L_(e)', 'ile_L_(e)', 'leu_L_(e)', 'lys_L_(e)', 'met_L_(e)', 'phe_L_(e)', \n",
    "              'pro_L_(e)', 'ser_L_(e)', 'thr_L_(e)', 'trp_L_(e)', 'tyr_L_(e)', 'val_L_(e)', 'glc_D_(e)', 'starch1200_(e)', \n",
    "              'cl_(e)', 'h_(e)',\"ac_(e)\", \"so4_(e)\" #'thiog_(e)'\n",
    "              , 'na1_(e)', 'fe3_(e)', 'ppp9_(e)', 'h2o_(e)', \n",
    "              'pi_(e)', 'k_(e)']\n",
    "MediumMgam=[mets[:-4] +\"(e)\" for mets in MediumMgam]\n",
    "print(MediumMgam)\n",
    "model2=model.copy()\n",
    "model_mgam=medium(model2,MediumMgam)\n",
    "print('\\nMedium_mgam')\n",
    "model_mgam.objective = {model_mgam.reactions.get_by_id('Biomass_FP'): 1}\n",
    "print(model_mgam.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ala_L(e)', 'arg_L(e)', 'asn_L(e)', 'asp_L(e)', 'asx_L(e)', 'cys_L(e)', 'glu_L(e)', 'gln_L(e)', 'glx_L(e)', 'gly_L(e)', 'his_L(e)', 'ile_L(e)', 'leu_L(e)', 'lys_L(e)', 'met_L(e)', 'phe_L(e)', 'pro_L(e)', 'ser_L(e)', 'thr_L(e)', 'trp_L(e)', 'tyr_L(e)', 'val_L(e)', 'glc_D(e)', 'starch1200(e)', 'cl(e)', 'h(e)', 'ac(e)', 'so4(e)', 'na1(e)', 'fe3(e)', 'ppp9(e)', 'h2o(e)', 'pi(e)', 'k(e)']\n",
      "\n",
      "Medium_mgam\n",
      "Objective\n",
      "=========\n",
      "1.0 Biomass_FP = 131.51398638093758\n",
      "\n",
      "Uptake\n",
      "------\n",
      " Metabolite       Reaction   Flux  C-Number  C-Flux\n",
      "    4hba[c]        DM_4HBA  64.77         7   4.63%\n",
      "    3mop[e]     EX_3mop(e)  521.6         6  31.94%\n",
      "   acglu[e]    EX_acglu(e)  57.79         7   4.13%\n",
      "     ade[e]      EX_ade(e)  28.42         5   1.45%\n",
      "  adocbl[e]   EX_adocbl(e) 0.7395        72   0.54%\n",
      "  alaala[e]   EX_alaala(e)  38.38         6   2.35%\n",
      "  alahis[e]   EX_alahis(e)  12.18         9   1.12%\n",
      "  alaleu[e]   EX_alaleu(e)  57.69         9   5.30%\n",
      "   arg-L[e]    EX_arg_L(e)  37.91         6   2.32%\n",
      "     btn[e]      EX_btn(e) 0.7395        10   0.08%\n",
      "     but[e]      EX_but(e)  27.58         4   1.13%\n",
      "     ca2[e]      EX_ca2(e) 0.7395         0   0.00%\n",
      "      cl[e]       EX_cl(e) 0.7395         0   0.00%\n",
      " cobalt2[e]  EX_cobalt2(e) 0.7395         0   0.00%\n",
      "  cps_fp[e]   EX_cps_fp(e)  27.35        63  17.59%\n",
      "     cu2[e]      EX_cu2(e) 0.7395         0   0.00%\n",
      "     fe2[e]      EX_fe2(e) 0.7395         0   0.00%\n",
      "     fe3[e]      EX_fe3(e) 0.7395         0   0.00%\n",
      "     fol[e]      EX_fol(e)  2.218        19   0.43%\n",
      "  glyc3p[e]   EX_glyc3p(e)   25.4         3   0.78%\n",
      "     gua[e]      EX_gua(e)  30.59         5   1.56%\n",
      "   ile-L[e]    EX_ile_L(e)  37.16         6   2.28%\n",
      "       k[e]        EX_k(e) 0.7395         0   0.00%\n",
      "   lys-L[e]    EX_lys_L(e)  56.03         6   3.43%\n",
      "  metala[e]   EX_metala(e)  20.38         8   1.66%\n",
      "     mg2[e]      EX_mg2(e) 0.7395         0   0.00%\n",
      "    ncam[e]     EX_ncam(e)  1.479         6   0.09%\n",
      "   phe-L[e]    EX_phe_L(e)   23.8         9   2.19%\n",
      "    plac[e]     EX_plac(e)  4.492         9   0.41%\n",
      "  pnto-R[e]   EX_pnto_R(e)  1.479         9   0.14%\n",
      "   pro-L[e]    EX_pro_L(e)  50.04         5   2.55%\n",
      "    pydx[e]     EX_pydx(e) 0.7395         8   0.06%\n",
      "  ribflv[e]   EX_ribflv(e)  1.479        17   0.26%\n",
      "     so4[e]      EX_so4(e) 0.7395         0   0.00%\n",
      "     thm[e]      EX_thm(e)  65.51        12   8.02%\n",
      "   trp-L[e]    EX_trp_L(e)  7.254        11   0.81%\n",
      "   val-L[e]    EX_val_L(e)  54.13         5   2.76%\n",
      "\n",
      "Secretion\n",
      "---------\n",
      "Metabolite    Reaction   Flux  C-Number  C-Flux\n",
      "   cgly[e]  EX_cgly(e) -24.01         5   7.96%\n",
      "    co2[e]   EX_co2(e) -229.9         1  15.24%\n",
      "    for[e]   EX_for(e) -475.6         1  31.54%\n",
      "    gal[e]   EX_gal(e) -26.61         6  10.59%\n",
      "    gam[e]   EX_gam(e) -2.288         6   0.91%\n",
      "  glc-D[e]   EX_glc(e) -53.22         6  21.18%\n",
      "    gly[e]   EX_gly(e) -24.86         2   3.30%\n",
      "    h2s[e]   EX_h2s(e) -27.58         0   0.00%\n",
      "     o2[e]    EX_o2(e) -32.34         0   0.00%\n",
      "   orot[e]  EX_orot(e) -10.49         5   3.48%\n",
      "    ura[e]   EX_ura(e) -21.91         4   5.81%\n",
      "\n"
     ]
    }
   ],
   "source": [
    "MediumMgam = ['ala_L_(e)', 'arg_L_(e)', 'asn_L_(e)', 'asp_L_(e)', 'asx_L_(e)', 'cys_L_(e)', 'glu_L_(e)', 'gln_L_(e)', \n",
    "              'glx_L_(e)', 'gly_L_(e)', 'his_L_(e)', 'ile_L_(e)', 'leu_L_(e)', 'lys_L_(e)', 'met_L_(e)', 'phe_L_(e)', \n",
    "              'pro_L_(e)', 'ser_L_(e)', 'thr_L_(e)', 'trp_L_(e)', 'tyr_L_(e)', 'val_L_(e)', 'glc_D_(e)', 'starch1200_(e)', \n",
    "              'cl_(e)', 'h_(e)',\"ac_(e)\", \"so4_(e)\" #'thiog_(e)'\n",
    "              , 'na1_(e)', 'fe3_(e)', 'ppp9_(e)', 'h2o_(e)', \n",
    "              'pi_(e)', 'k_(e)']\n",
    "MediumMgam=[mets[:-4] +\"(e)\" for mets in MediumMgam]\n",
    "print(MediumMgam)\n",
    "model2=model.copy()\n",
    "model_mgam=medium(model2,MediumMgam)\n",
    "print('\\nMedium_mgam')\n",
    "for i in range(1,len(model_mgam.reactions)):\n",
    "    model_mgam.reactions[i].upper_bound=1000\n",
    "    model_mgam.reactions[i].lower_bound=-1000 #does not change till 77,78\n",
    "model_mgam.objective = {model_mgam.reactions.get_by_id('Biomass_FP'): 1}\n",
    "print(model_mgam.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Active reaction : 364 which represent 35.339806% of the model\n",
      "69.8788559078775\n"
     ]
    }
   ],
   "source": [
    "fba = model.optimize()\n",
    "fluxes = deleteNull(fba.fluxes)\n",
    "print('\\nActive reaction : %i which represent %f%% of the model' %(len(fluxes),len(fluxes)/len(model.reactions)*100))\n",
    "print(fba.objective_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Active reaction : 381 which represent 36.990291% of the model\n",
      "131.5139863809376\n"
     ]
    }
   ],
   "source": [
    "fba_model_mgam=model_mgam.optimize()\n",
    "fluxes_model_mgam = deleteNull(fba_model_mgam.fluxes)\n",
    "print('\\nActive reaction : %i which represent %f%% of the model' %(len(fluxes_model_mgam),len(fluxes_model_mgam)/len(model_mgam.reactions)*100))\n",
    "print(fba_model_mgam.objective_value)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reinforced Clostridial Medium (RCM)\n",
    "http://www.oxoid.com/UK/blue/prod_detail/prod_detail.asp?pr=CM0149&org=53&c=UK&lang=EN\n",
    "\n",
    "- Yeast Extract -> lots of stuff in it, unsure of what's garbage\n",
    "\n",
    "    - Glutathione: gthrd (reduced version)\n",
    "    - 18 AA's: for concentration see MediumMainComponents\n",
    "    - Dextran: glc (it's made of glucose)\n",
    "    - Mannan: mannan or mannose -> man\n",
    "    - Trehalose: tre (not present in exchange reactions)\n",
    "    - B-vitamins: (Clemence mentioned vitamins being important so I'm putting them all in)\n",
    "        - Thiamin -> thm\n",
    "        - Riboflavin -> ribflv\n",
    "        - Niacin -> trp_L (made of tryptophan)\n",
    "        - Pantothenic Acid -> pnto_R\n",
    "        - Pyridoxine -> pydxn\n",
    "        - Biotin -> btn\n",
    "        - Folic Acid -> folate -> fol\n",
    "        - Cobalamin -> b12 / cbl1\n",
    "    - Biotin\n",
    "- Peptone -> Source of AA. We stick to bacteriological peptone\n",
    "- Glucose -> Assuming D_Glucose: glc__D\n",
    "- Soluble Starch -> strch1 or strch2\n",
    "- Sodium Chloride -> na1, cl\n",
    "- Sodium Acetate -> ac, na1\n",
    "- Cysteine Hydrochloride -> cys_L, cl, h (could also be cys_D? cys_L is more widely used)\n",
    "- Agar: Agarose (70%) and Agaropectin (30%) according to Wiki\n",
    "    - Agarose: D-galactose & 3,6-anhydro-L-galactopyranose\n",
    "        - D-galactose: gal\n",
    "        - 3,6-anhydro-L-galactopyranose: similar structure & compositon to beta-D-allose. D-Allose: all__D\n",
    "    - Agaropectin: D-galactose, L-galactose, pyruvate, sulfate\n",
    "        - D-galactose: gal\n",
    "        - L-galactose: gal__L\n",
    "        - Pyruvate: pyr (not present in exchange reactions)\n",
    "        - Sulfate: so4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<Reaction EX_3mop(e) at 0x21f410e8a60>,\n",
       " <Reaction EX_4abz(e) at 0x21f410e8af0>,\n",
       " <Reaction EX_5oxpro(e) at 0x21f410e8a30>,\n",
       " <Reaction EX_Lcyst(e) at 0x21f410e8a90>,\n",
       " <Reaction EX_Lcystin(e) at 0x21f410e8ac0>,\n",
       " <Reaction EX_ac(e) at 0x21f410e8b80>,\n",
       " <Reaction EX_acasp(e) at 0x21f410e8bb0>,\n",
       " <Reaction EX_acgal(e) at 0x21f410e8be0>,\n",
       " <Reaction EX_acgalglcur(e) at 0x21f410e8c10>,\n",
       " <Reaction EX_acgam(e) at 0x21f410e8c40>,\n",
       " <Reaction EX_acglu(e) at 0x21f410e8c70>,\n",
       " <Reaction EX_acnam(e) at 0x21f410e8ca0>,\n",
       " <Reaction EX_ade(e) at 0x21f410e8cd0>,\n",
       " <Reaction EX_adocbl(e) at 0x21f410e8d00>,\n",
       " <Reaction EX_ala_D(e) at 0x21f410e8d30>,\n",
       " <Reaction EX_ala_L(e) at 0x21f410e8b20>,\n",
       " <Reaction EX_alaala(e) at 0x21f410e8d60>,\n",
       " <Reaction EX_alaasp(e) at 0x21f410e8d90>,\n",
       " <Reaction EX_alagln(e) at 0x21f410e87f0>,\n",
       " <Reaction EX_alaglu(e) at 0x21f410e8dc0>,\n",
       " <Reaction EX_alagly(e) at 0x21f410e8df0>,\n",
       " <Reaction EX_alahis(e) at 0x21f410e8e20>,\n",
       " <Reaction EX_alaleu(e) at 0x21f410e8e50>,\n",
       " <Reaction EX_alathr(e) at 0x21f410e8e80>,\n",
       " <Reaction EX_arab_L(e) at 0x21f410e8eb0>,\n",
       " <Reaction EX_arg_L(e) at 0x21f410e8ee0>,\n",
       " <Reaction EX_asn_L(e) at 0x21f410e8f10>,\n",
       " <Reaction EX_aso3(e) at 0x21f410e8a00>,\n",
       " <Reaction EX_aso4(e) at 0x21f410e8f40>,\n",
       " <Reaction EX_asp_L(e) at 0x21f410e8f70>,\n",
       " <Reaction EX_btn(e) at 0x21f410e8fa0>,\n",
       " <Reaction EX_but(e) at 0x21f410e8fd0>,\n",
       " <Reaction EX_butso3(e) at 0x21f410fa040>,\n",
       " <Reaction EX_ca2(e) at 0x21f410fa070>,\n",
       " <Reaction EX_cbl1(e) at 0x21f410fa0a0>,\n",
       " <Reaction EX_cbl2(e) at 0x21f410fa0d0>,\n",
       " <Reaction EX_cd2(e) at 0x21f410fa100>,\n",
       " <Reaction EX_cellb(e) at 0x21f410fa130>,\n",
       " <Reaction EX_cgly(e) at 0x21f410fa160>,\n",
       " <Reaction EX_cit(e) at 0x21f410fa190>,\n",
       " <Reaction EX_cl(e) at 0x21f410fa1c0>,\n",
       " <Reaction EX_co2(e) at 0x21f410fa1f0>,\n",
       " <Reaction EX_cobalt2(e) at 0x21f410fa220>,\n",
       " <Reaction EX_cps_fp(e) at 0x21f410fa250>,\n",
       " <Reaction EX_cu2(e) at 0x21f410fa280>,\n",
       " <Reaction EX_cys_L(e) at 0x21f410fa2b0>,\n",
       " <Reaction EX_dhor_S(e) at 0x21f410fa2e0>,\n",
       " <Reaction EX_ethso3(e) at 0x21f410fa310>,\n",
       " <Reaction EX_fe2(e) at 0x21f410fa340>,\n",
       " <Reaction EX_fe3(e) at 0x21f410fa370>,\n",
       " <Reaction EX_fe3dcit(e) at 0x21f410fa3a0>,\n",
       " <Reaction EX_fol(e) at 0x21f410fa3d0>,\n",
       " <Reaction EX_for(e) at 0x21f410fa400>,\n",
       " <Reaction EX_fru(e) at 0x21f410fa430>,\n",
       " <Reaction EX_fum(e) at 0x21f410fa460>,\n",
       " <Reaction EX_gal(e) at 0x21f410fa490>,\n",
       " <Reaction EX_galur(e) at 0x21f410fa4c0>,\n",
       " <Reaction EX_gam(e) at 0x21f410fa4f0>,\n",
       " <Reaction EX_glc(e) at 0x21f410fa520>,\n",
       " <Reaction EX_glcur(e) at 0x21f410fa550>,\n",
       " <Reaction EX_gln_L(e) at 0x21f410fa580>,\n",
       " <Reaction EX_glu_L(e) at 0x21f410fa5b0>,\n",
       " <Reaction EX_gly(e) at 0x21f410fa5e0>,\n",
       " <Reaction EX_glyasn(e) at 0x21f410fa610>,\n",
       " <Reaction EX_glyasp(e) at 0x21f410fa640>,\n",
       " <Reaction EX_glyc3p(e) at 0x21f410fa670>,\n",
       " <Reaction EX_glyc_R(e) at 0x21f410fa6a0>,\n",
       " <Reaction EX_glygln(e) at 0x21f410fa6d0>,\n",
       " <Reaction EX_glyglu(e) at 0x21f410fa700>,\n",
       " <Reaction EX_glygly(e) at 0x21f410fa730>,\n",
       " <Reaction EX_glyleu(e) at 0x21f410fa760>,\n",
       " <Reaction EX_glymet(e) at 0x21f410fa790>,\n",
       " <Reaction EX_glyphe(e) at 0x21f410fa7c0>,\n",
       " <Reaction EX_glypro(e) at 0x21f410fa7f0>,\n",
       " <Reaction EX_glytyr(e) at 0x21f410fa820>,\n",
       " <Reaction EX_gthox(e) at 0x21f410fa850>,\n",
       " <Reaction EX_gthrd(e) at 0x21f410fa880>,\n",
       " <Reaction EX_gua(e) at 0x21f410fa8b0>,\n",
       " <Reaction EX_h(e) at 0x21f410fa8e0>,\n",
       " <Reaction EX_h2o(e) at 0x21f410fa910>,\n",
       " <Reaction EX_h2s(e) at 0x21f410fa940>,\n",
       " <Reaction EX_hg2(e) at 0x21f410fa970>,\n",
       " <Reaction EX_his_L(e) at 0x21f410fa9a0>,\n",
       " <Reaction EX_hxan(e) at 0x21f410fa9d0>,\n",
       " <Reaction EX_ile_L(e) at 0x21f410faa00>,\n",
       " <Reaction EX_ins(e) at 0x21f410faa30>,\n",
       " <Reaction EX_inulin(e) at 0x21f410faa60>,\n",
       " <Reaction EX_isetac(e) at 0x21f410faa90>,\n",
       " <Reaction EX_k(e) at 0x21f410faac0>,\n",
       " <Reaction EX_kesto(e) at 0x21f410faaf0>,\n",
       " <Reaction EX_kestopt(e) at 0x21f410fab20>,\n",
       " <Reaction EX_kestottr(e) at 0x21f410fab50>,\n",
       " <Reaction EX_lac_D(e) at 0x21f410fab80>,\n",
       " <Reaction EX_lcts(e) at 0x21f410fabb0>,\n",
       " <Reaction EX_leu_L(e) at 0x21f410fabe0>,\n",
       " <Reaction EX_leugly(e) at 0x21f410fac10>,\n",
       " <Reaction EX_leuleu(e) at 0x21f410fac40>,\n",
       " <Reaction EX_lys_L(e) at 0x21f410fac70>,\n",
       " <Reaction EX_mal_L(e) at 0x21f410faca0>,\n",
       " <Reaction EX_malt(e) at 0x21f410facd0>,\n",
       " <Reaction EX_man(e) at 0x21f410fad00>,\n",
       " <Reaction EX_meoh(e) at 0x21f410fad30>,\n",
       " <Reaction EX_met_D(e) at 0x21f410fad60>,\n",
       " <Reaction EX_met_L(e) at 0x21f410fad90>,\n",
       " <Reaction EX_metala(e) at 0x21f410fadc0>,\n",
       " <Reaction EX_metsox_R_L(e) at 0x21f410fadf0>,\n",
       " <Reaction EX_metsox_S_L(e) at 0x21f410fae20>,\n",
       " <Reaction EX_mg2(e) at 0x21f410fae50>,\n",
       " <Reaction EX_mobd(e) at 0x21f410fae80>,\n",
       " <Reaction EX_mso3(e) at 0x21f410faeb0>,\n",
       " <Reaction EX_na1(e) at 0x21f410faee0>,\n",
       " <Reaction EX_nac(e) at 0x21f410faf10>,\n",
       " <Reaction EX_ncam(e) at 0x21f410faf40>,\n",
       " <Reaction EX_nh4(e) at 0x21f410faf70>,\n",
       " <Reaction EX_o2(e) at 0x21f410fafa0>,\n",
       " <Reaction EX_orn(e) at 0x21f410fafd0>,\n",
       " <Reaction EX_orot(e) at 0x21f41110040>,\n",
       " <Reaction EX_pb(e) at 0x21f41110070>,\n",
       " <Reaction EX_pect(e) at 0x21f411100a0>,\n",
       " <Reaction EX_phe_L(e) at 0x21f411100d0>,\n",
       " <Reaction EX_pi(e) at 0x21f41110100>,\n",
       " <Reaction EX_plac(e) at 0x21f41110130>,\n",
       " <Reaction EX_pnto_R(e) at 0x21f41110160>,\n",
       " <Reaction EX_pro_L(e) at 0x21f41110190>,\n",
       " <Reaction EX_progly(e) at 0x21f411101c0>,\n",
       " <Reaction EX_ptrc(e) at 0x21f411101f0>,\n",
       " <Reaction EX_pydam(e) at 0x21f41110220>,\n",
       " <Reaction EX_pydx(e) at 0x21f41110250>,\n",
       " <Reaction EX_pydxn(e) at 0x21f41110280>,\n",
       " <Reaction EX_rbflvrd(e) at 0x21f411102b0>,\n",
       " <Reaction EX_ribflv(e) at 0x21f411102e0>,\n",
       " <Reaction EX_seln(e) at 0x21f41110310>,\n",
       " <Reaction EX_ser_L(e) at 0x21f41110340>,\n",
       " <Reaction EX_so4(e) at 0x21f41110370>,\n",
       " <Reaction EX_spmd(e) at 0x21f411103a0>,\n",
       " <Reaction EX_strch1(e) at 0x21f411103d0>,\n",
       " <Reaction EX_succ(e) at 0x21f41110400>,\n",
       " <Reaction EX_sulfac(e) at 0x21f41110430>,\n",
       " <Reaction EX_taur(e) at 0x21f41110460>,\n",
       " <Reaction EX_thm(e) at 0x21f41110490>,\n",
       " <Reaction EX_thr_L(e) at 0x21f411104c0>,\n",
       " <Reaction EX_thymd(e) at 0x21f411104f0>,\n",
       " <Reaction EX_trp_L(e) at 0x21f41110520>,\n",
       " <Reaction EX_tyr_L(e) at 0x21f41110550>,\n",
       " <Reaction EX_ura(e) at 0x21f41110580>,\n",
       " <Reaction EX_urate(e) at 0x21f411105b0>,\n",
       " <Reaction EX_urea(e) at 0x21f411105e0>,\n",
       " <Reaction EX_val_L(e) at 0x21f41110610>,\n",
       " <Reaction EX_xan(e) at 0x21f41110640>,\n",
       " <Reaction EX_xyl_D(e) at 0x21f41110670>]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.exchanges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Objective\n",
      "=========\n",
      "1.0 Biomass_FP = 69.8788559078775\n",
      "\n",
      "Uptake\n",
      "------\n",
      "    Metabolite          Reaction   Flux  C-Number  C-Flux\n",
      "    Lcystin[e]     EX_Lcystin(e)  3.502         6   0.05%\n",
      "        ade[e]         EX_ade(e)  864.9         5  10.85%\n",
      "     adocbl[e]      EX_adocbl(e) 0.3929        72   0.07%\n",
      "     alaala[e]      EX_alaala(e)  6.462         6   0.10%\n",
      "     alagln[e]      EX_alagln(e)  30.81         8   0.62%\n",
      "      arg-L[e]       EX_arg_L(e)   1000         6  15.05%\n",
      "      asn-L[e]       EX_asn_L(e)   16.4         4   0.16%\n",
      "        btn[e]         EX_btn(e) 0.3929        10   0.01%\n",
      "        ca2[e]         EX_ca2(e) 0.3929         0   0.00%\n",
      "         cl[e]          EX_cl(e) 0.3929         0   0.00%\n",
      "    cobalt2[e]     EX_cobalt2(e) 0.3929         0   0.00%\n",
      "     cps_fp[e]      EX_cps_fp(e) 0.3929        63   0.06%\n",
      "        cu2[e]         EX_cu2(e) 0.3929         0   0.00%\n",
      "        fe3[e]         EX_fe3(e) 0.7858         0   0.00%\n",
      "        fol[e]         EX_fol(e)  1.179        19   0.06%\n",
      "        fum[e]         EX_fum(e)   1000         4  10.03%\n",
      "      glu-L[e]       EX_glu_L(e)   1000         5  12.54%\n",
      "        gua[e]         EX_gua(e)  16.25         5   0.20%\n",
      "          h[e]           EX_h(e)    901         0   0.00%\n",
      "      his-L[e]       EX_his_L(e)  6.472         6   0.10%\n",
      "        ins[e]         EX_ins(e)  751.8        10  18.86%\n",
      "          k[e]           EX_k(e) 0.3929         0   0.00%\n",
      "     leuleu[e]      EX_leuleu(e)  15.33        12   0.46%\n",
      "      lys-L[e]       EX_lys_L(e)  23.31         6   0.35%\n",
      "      mal-L[e]       EX_mal_L(e)   1000         4  10.03%\n",
      "     metala[e]      EX_metala(e)  10.59         8   0.21%\n",
      " metsox-S-L[e]  EX_metsox_S_L(e) 0.2368         5   0.00%\n",
      "        mg2[e]         EX_mg2(e) 0.3929         0   0.00%\n",
      "        na1[e]         EX_na1(e)   1000         0   0.00%\n",
      "        nac[e]         EX_nac(e) 0.7858         6   0.01%\n",
      "         o2[e]          EX_o2(e)  618.9         0   0.00%\n",
      "      phe-L[e]       EX_phe_L(e)  12.65         9   0.29%\n",
      "         pi[e]          EX_pi(e)  126.6         0   0.00%\n",
      "     pnto-R[e]      EX_pnto_R(e) 0.7858         9   0.02%\n",
      "      pro-L[e]       EX_pro_L(e)  15.05         5   0.19%\n",
      "       pydx[e]        EX_pydx(e) 0.3929         8   0.01%\n",
      "     ribflv[e]      EX_ribflv(e)  148.3        17   6.32%\n",
      "      ser-L[e]       EX_ser_L(e)  223.1         3   1.68%\n",
      "        so4[e]         EX_so4(e) 0.3929         0   0.00%\n",
      "        thm[e]         EX_thm(e) 0.3929        12   0.01%\n",
      "      thr-L[e]       EX_thr_L(e)  606.4         4   6.08%\n",
      "      thymd[e]       EX_thymd(e)  1.642        10   0.04%\n",
      "      trp-L[e]       EX_trp_L(e)  3.854        11   0.11%\n",
      "      tyr-L[e]       EX_tyr_L(e)  9.381         9   0.21%\n",
      "        ura[e]         EX_ura(e)  20.52         4   0.21%\n",
      "      urate[e]       EX_urate(e)  398.4         5   5.00%\n",
      "       hpyr[c]         sink_hpyr   0.05         3   0.00%\n",
      "\n",
      "Secretion\n",
      "---------\n",
      " Metabolite       Reaction   Flux  C-Number  C-Flux\n",
      "     na1[c]         DM_NA1  -1000         0   0.00%\n",
      "      ac[e]       EX_ac(e)  -1000         2   5.64%\n",
      "   asp-L[e]    EX_asp_L(e) -977.1         4  11.02%\n",
      "     cit[e]      EX_cit(e) -988.6         6  16.73%\n",
      "     co2[e]      EX_co2(e)  -1000         1   2.82%\n",
      "  dhor-S[e]   EX_dhor_S(e) -785.4         5  11.07%\n",
      "     for[e]      EX_for(e)   -158         1   0.45%\n",
      "     gly[e]      EX_gly(e) -527.8         2   2.98%\n",
      "  glyc-R[e]   EX_glyc_R(e)  -0.05         3   0.00%\n",
      "    hxan[e]     EX_hxan(e)  -1000         5  14.10%\n",
      "     nh4[e]      EX_nh4(e)  -1000         0   0.00%\n",
      "    orot[e]     EX_orot(e) -194.5         5   2.74%\n",
      "    ptrc[e]     EX_ptrc(e) -979.9         4  11.05%\n",
      " rbflvrd[e]  EX_rbflvrd(e) -147.5        17   7.07%\n",
      "    succ[e]     EX_succ(e) -20.14         4   0.23%\n",
      "     xan[e]      EX_xan(e)  -1000         5  14.10%\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(model.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Objective\n",
      "=========\n",
      "1.0 Biomass_FP = 30.73892495631308\n",
      "\n",
      "Uptake\n",
      "------\n",
      " Metabolite       Reaction   Flux  C-Number  C-Flux\n",
      " Lcystin[e]  EX_Lcystin(e)   1.54         6   0.10%\n",
      "     ade[e]      EX_ade(e)    100         5   5.34%\n",
      "  adocbl[e]   EX_adocbl(e) 0.1728        72   0.13%\n",
      "   ala-D[e]    EX_ala_D(e)  5.685         3   0.18%\n",
      "   ala-L[e]    EX_ala_L(e)  18.21         3   0.58%\n",
      "   arg-L[e]    EX_arg_L(e)   15.5         6   0.99%\n",
      "   asn-L[e]    EX_asn_L(e)  7.214         4   0.31%\n",
      "     btn[e]      EX_btn(e) 0.1728        10   0.02%\n",
      "     ca2[e]      EX_ca2(e) 0.1728         0   0.00%\n",
      "      cl[e]       EX_cl(e) 0.1728         0   0.00%\n",
      " cobalt2[e]  EX_cobalt2(e) 0.1728         0   0.00%\n",
      "     cu2[e]      EX_cu2(e) 0.1728         0   0.00%\n",
      "     fe2[e]      EX_fe2(e) 0.1728         0   0.00%\n",
      "     fe3[e]      EX_fe3(e) 0.1728         0   0.00%\n",
      "     fol[e]      EX_fol(e) 0.5185        19   0.11%\n",
      "   galur[e]    EX_galur(e)    100         6   6.41%\n",
      "   glc-D[e]      EX_glc(e)    100         6   6.41%\n",
      "   gln-L[e]    EX_gln_L(e)  13.73         5   0.73%\n",
      "   glu-L[e]    EX_glu_L(e)  4.854         5   0.26%\n",
      "     gua[e]      EX_gua(e)   7.15         5   0.38%\n",
      "     h2o[e]      EX_h2o(e)    100         0   0.00%\n",
      "   his-L[e]    EX_his_L(e)  2.847         6   0.18%\n",
      "     ins[e]      EX_ins(e)    100        10  10.68%\n",
      "       k[e]        EX_k(e) 0.1728         0   0.00%\n",
      "   leu-L[e]    EX_leu_L(e)  13.48         6   0.86%\n",
      "   lys-L[e]    EX_lys_L(e)  10.25         6   0.66%\n",
      "   met-L[e]    EX_met_L(e)  4.763         5   0.25%\n",
      "     mg2[e]      EX_mg2(e) 0.1728         0   0.00%\n",
      "     na1[e]      EX_na1(e)    100         0   0.00%\n",
      "     nac[e]      EX_nac(e) 0.3457         6   0.02%\n",
      "    pect[e]     EX_pect(e)  1.659      2535  44.91%\n",
      "   phe-L[e]    EX_phe_L(e)  5.564         9   0.53%\n",
      "      pi[e]       EX_pi(e)  57.08         0   0.00%\n",
      "  pnto-R[e]   EX_pnto_R(e) 0.3457         9   0.03%\n",
      "   pro-L[e]    EX_pro_L(e)  6.622         5   0.35%\n",
      "    pydx[e]     EX_pydx(e) 0.1728         8   0.01%\n",
      "  ribflv[e]   EX_ribflv(e)  83.09        17  15.09%\n",
      "   ser-L[e]    EX_ser_L(e)  57.84         3   1.85%\n",
      "     so4[e]      EX_so4(e) 0.1728         0   0.00%\n",
      "     thm[e]      EX_thm(e) 0.1728        12   0.02%\n",
      "   thr-L[e]    EX_thr_L(e)  34.58         4   1.48%\n",
      "   thymd[e]    EX_thymd(e) 0.8953        10   0.10%\n",
      "   trp-L[e]    EX_trp_L(e)  1.695        11   0.20%\n",
      "   tyr-L[e]    EX_tyr_L(e)  4.127         9   0.40%\n",
      "     ura[e]      EX_ura(e)  9.545         4   0.41%\n",
      "    hpyr[c]      sink_hpyr   0.05         3   0.00%\n",
      "\n",
      "Secretion\n",
      "---------\n",
      " Metabolite       Reaction   Flux  C-Number  C-Flux\n",
      "     na1[c]         DM_NA1   -100         0   0.00%\n",
      "      ac[e]       EX_ac(e) -852.2         2  22.95%\n",
      "   asp-L[e]    EX_asp_L(e) -114.1         4   6.15%\n",
      "     co2[e]      EX_co2(e)  -1000         1  13.47%\n",
      "  dhor-S[e]   EX_dhor_S(e) -6.643         5   0.45%\n",
      "     for[e]      EX_for(e) -33.66         1   0.45%\n",
      "  glyc-R[e]   EX_glyc_R(e)  -0.05         3   0.00%\n",
      "       h[e]        EX_h(e) -708.9         0   0.00%\n",
      "    hxan[e]     EX_hxan(e) -193.4         5  13.02%\n",
      "   lac-D[e]    EX_lac_D(e) -491.2         3  19.85%\n",
      "    meoh[e]     EX_meoh(e) -323.4         1   4.36%\n",
      "    ptrc[e]     EX_ptrc(e) -6.643         4   0.36%\n",
      " rbflvrd[e]  EX_rbflvrd(e) -82.75        17  18.94%\n",
      "\n"
     ]
    }
   ],
   "source": [
    "medium_rcm = ['4abz(e)','5oxpro(e)','ac(e)',' Lcyst(e)', 'Lcystin(e)','ade(e)', 'adocbl(e)',\n",
    "              'ala_D(e)', 'ala_L(e)', 'arg_L(e)','asn_L(e)', 'asp_L(e)', 'btn(e)', 'ca2(e)', \n",
    "              'cbl1(e)','cbl2(e)','cl(e)','cobalt2(e)','cys_L(e)', 'cu2(e)','fe2(e)', 'fe3(e)',\n",
    "              'fol(e)', 'gal(e)','galur(e)','glc(e)','gln_L(e)', 'glu_L(e)', 'gly(e)', \n",
    "              'gthox(e)','gthrd(e)', 'gua(e)', 'h(e)','h2o(e)','his_L(e)','hxan(e)',\n",
    "              'ile_L(e)','ins(e)', 'k(e)','leu_L(e)', 'lys_L(e)','man(e)', 'mg2(e)',\n",
    "              'met_D(e)', 'met_L(e)','na1(e)', 'nac(e)', 'ncam(e)','pb(e)','pect(e)', 'phe_L(e)','pi(e)',\n",
    "              'pnto_R(e)', 'ptrc(e)', 'pro_L(e)', 'pydam(e)','pydx(e)', 'pydxn(e)', \n",
    "              'rbflvrd(e)', 'ribflv(e)', 'ser_L(e)', 'so4(e)','spmd(e)',\n",
    "              'strch1(e)',  'taur(e)', 'thm(e)', 'thr_L(e)', 'thymd(e)', 'trp_L(e)', \n",
    "              'tyr_L(e)', 'ura(e)', 'val_L(e)', 'xan(e)']\n",
    "# missing: tre, all_D, pyr.\n",
    "model_rcm = medium(model.copy(), medium_rcm)\n",
    "print(model_rcm.summary())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EX_4abz(e): 4abz[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_5oxpro(e): 5oxpro[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_Lcystin(e): Lcystin[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_ac(e): ac[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_ade(e): ade[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_adocbl(e): adocbl[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_ala_D(e): ala-D[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_ala_L(e): ala-L[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_arg_L(e): arg-L[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_asn_L(e): asn-L[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_asp_L(e): asp-L[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_btn(e): btn[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_ca2(e): ca2[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_cbl1(e): cbl1[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_cbl2(e): cbl2[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_cl(e): cl[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_cobalt2(e): cobalt2[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_cu2(e): cu2[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_cys_L(e): cys-L[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_fe2(e): fe2[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_fe3(e): fe3[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_fol(e): fol[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_gal(e): gal[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_galur(e): galur[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_glc(e): glc-D[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_gln_L(e): gln-L[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_glu_L(e): glu-L[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_gly(e): gly[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_gthox(e): gthox[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_gthrd(e): gthrd[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_gua(e): gua[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_h(e): h[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_h2o(e): h2o[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_his_L(e): his-L[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_hxan(e): hxan[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_ile_L(e): ile-L[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_ins(e): ins[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_k(e): k[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_leu_L(e): leu-L[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_lys_L(e): lys-L[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_man(e): man[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_met_D(e): met-D[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_met_L(e): met-L[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_mg2(e): mg2[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_na1(e): na1[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_nac(e): nac[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_ncam(e): ncam[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_pb(e): pb[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_pect(e): pect[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_pi(e): pi[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_pnto_R(e): pnto-R[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_pro_L(e): pro-L[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_ptrc(e): ptrc[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_pydam(e): pydam[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_pydx(e): pydx[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_pydxn(e): pydxn[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_rbflvrd(e): rbflvrd[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_ribflv(e): ribflv[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_ser_L(e): ser-L[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_so4(e): so4[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_spmd(e): spmd[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_strch1(e): strch1[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_taur(e): taur[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_thm(e): thm[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_thr_L(e): thr-L[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_thymd(e): thymd[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_trp_L(e): trp-L[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_tyr_L(e): tyr-L[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_ura(e): ura[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_val_L(e): val-L[e] <=> \n",
      "1000.0\n",
      "-100\n",
      "EX_xan(e): xan[e] <=> \n",
      "1000.0\n",
      "-100\n"
     ]
    }
   ],
   "source": [
    "for i in range(1,len(model_rcm.reactions)):\n",
    "    if model_rcm.reactions[i].lower_bound == -100:\n",
    "        print(model.reactions[i])\n",
    "        print(model_rcm.reactions[i].upper_bound)\n",
    "        print(model_rcm.reactions[i].lower_bound)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
